### ISSUE TITLE
amp has problems with mixed case keyword args

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#1991]
|<img alt="scmikes's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| scmikes reported|
|-|-|
|Trac ID|trac#1991|
|Type|defect|
|Created|2006-08-08 04:44:39Z|

Not sure if this is a bug or feature.

Attached is a very simple amp client and server.

This version fails.  If you change fileName to fileName.  Then all is well

Attachments:

* [fileTransferAmpServer.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/643/643fec50e79c69bc6bbb7616afd3904acf40867c/bbfcef492a4d8353cf6625c8771e7ffb25a17236.py) (696 bytes) - added by scmikes on 2006-08-08 04:45:26Z - 
* [fileTransferAmpServer.2.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/643/643fec50e79c69bc6bbb7616afd3904acf40867c/3ef3eba09f71fd590f5a3d60a4fab3ad9a9a6e61.py) (696 bytes) - added by scmikes on 2006-08-08 04:45:42Z - 
* [fileTransferClient.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/643/643fec50e79c69bc6bbb7616afd3904acf40867c/c5951db71ab34ed6d4c425b884ebb4e69109b77c.py) (1232 bytes) - added by scmikes on 2006-08-08 04:46:15Z - 
* [fileTransferAmpServer.3.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/643/643fec50e79c69bc6bbb7616afd3904acf40867c/fdde09ab250d01a241a364bd63a872cc8e2c5c8e.py) (696 bytes) - added by scmikes on 2006-08-08 05:05:37Z - working server (no camel case) filename
* [fileTransferClient.2.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/643/643fec50e79c69bc6bbb7616afd3904acf40867c/6dcefc1aa864243add41b4a05f827847b236c819.py) (1232 bytes) - added by scmikes on 2006-08-08 05:06:23Z - working amp client (no camel case filename)

<details><summary>Searchable metadata</summary>

```
trac-id__1991 1991
type__defect defect
reporter__scmikes scmikes
priority__high high
milestone__Core_2_5 Core-2.5
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__amp amp
time__1155012279000000 1155012279000000
changetime__1155709353000000 1155709353000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|<img alt="scmikes's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_1"></a>|scmikes commented|
|-|-|

Sorry for the duplicate AmpServer.py (2nd file)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_3"></a>|@glyph commented|
|-|-|

There was a tiny feature here buried under a pile of crap.

I've written some tests, fixed the bug, made the name-normalization private and documented its purpose a bit better, in ```ampquote-1991```.  I think it's ready for review.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_4"></a>|@glyph commented|
|-|-|

To explain a bit further: the reason it was doing all this stuff with lowercasing was that Juice went through some ridiculous gyrations to totally "normalize" the format of all keys.  This is no longer the case, because they don't have to fit into the rigid syntax of ": "-separated pairs.  In the branch I mentioned, it only "normalizes" keys to turn names from the wire into valid Python identifiers.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_5"></a>|@glyph commented|
|-|-|

Jerub cannot access the website for some reason, there are connectivity problems to .au right now.  But he checked out the code before those problems, and reviewed it, and can still get to IRC just fine - he assures me he has reviewed it.

He also couldn't find problems:

```
(01:41:50) Jerub: "kthnxbai, merge plz"
(01:44:36) Jerub: but seriously, it's a trivial change with tests, for behaviour that 
    shouldn't have been broken in the first place.
```

So I'm going to merge it.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_6"></a>|@glyph set status to `closed`|
|-|-|

(In [17840]) Remove vestiges of JUICE-style header quoting in AMP.

Among other things, this makes mixed-case argument and command names work.
Previously they would raise exceptions and kill the connection if you tried to
use them.  Although the amount of key-mangling has been significantly reduced,
I've included a detailed explanation of why it's being done at all any more,
where it is.

This also fixes a minor cleanup/pyflakes issue with test_amp; it wasn't causing
problems, but it seemed like a likely place for a race-condition to pop up.

Author: glyph

Reviewer: jerub

Fixes [#1991](https://github.com/twisted/twisted/issues/1991)

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
